source/main/utils/handleCheckBlockReplayProgress.ts

```diff
@@ -65,15 +68,26 @@ const createHandleNewLogLine = (mainWindow: BrowserWindow) => {
       return;
     }
 
-    const progress = Math.floor(parseFloat(unparsedProgress));
+    // In rare cases cardano-node does not log 100%, therefore we need to manually mark the previous step as complete.
```

Comment:
This mitigates the risk of having a race condition. In rare cases the log entries can be pushed before we set up the file watch mechanism, so this additional logic ensures we mark previous steps as complete when the next steps starts. This also improves UX as currently sometimes cardano-node doesn't report 100% progress.

Reply:
Alternatively we can read past entries from the file. @szymonmaslowski found out that there is a param in `tail` that allows us to do that.

But to be fair, I believe it just depends how you interpret logs. Eg. you can interpret the following entry:

```
[34m[Marcins-:cardano.node.ChainDB:Info:19][0m [2022-04-27 12:30:36.94 UTC] Opened lgr db
[34m[Marcins-:cardano.node.ChainDB:Info:19][0m [2022-04-27 12:30:36.94 UTC] Started initial chain selection
```

as "replayed block is completed". It just doesn't match the `Progress: x%`. But makes the UX much better and shows the truth. The steps are not sequential but previous logic would often result in showing the previous step is stuck at eg. 67%, while that's not true.

Reply:
Are we 100% the steps are sequential? Often cardano-node would output in a different order depending of the ledger state.

Reply:
Although I observed this behaviour consistently while testing, I might be wrong. Since this implementation is controversial, I'll reimplement this with the approach @szymonmaslowski suggested - using the `fromBeginning` param from `tail`.